# Guided Project: Exploring eBay Car Sales Data

The data dictionary provided with data is as follows:
<ol>
<li>dateCrawled - When this ad was first crawled. All field-values are taken from this date.</li>
<li>name - Name of the car.</li>
<li>seller - Whether the seller is private or a dealer.</li>
<li>offerType - The type of listing</li>
<li>price - The price on the ad to sell the car.</li>
<li>abtest - Whether the listing is included in an A/B test.</li>
<li>vehicleType - The vehicle Type.</li>
<li>yearOfRegistration - The year in which which year the car was first registered.</li>
<li>gearbox - The transmission type.</li>
<li>powerPS - The power of the car in PS.</li>
<li>model - The car model name.</li>
<li>kilometer - How many kilometers the car has driven.</li>
<li>monthOfRegistration - The month in which which year the car was first registered.</li>
<li>fuelType - What type of fuel the car uses.</li>
<li>brand - The brand of the car.</li>
<li>notRepairedDamage - If the car has a damage which is not yet repaired.</li>
<li>dateCreated - The date on which the eBay listing was created.</li>
<li>nrOfPictures - The number of pictures in the ad.</li>
<li>postalCode - The postal code for the location of the vehicle.</li>
<li>lastSeenOnline - When the crawler saw this ad last online.</li>


In [1]:
#Import pandas and NumPy libraries
import pandas as pd
import numpy as np

#Read in a CSV file into pandas with a specific coding, and assign it to the variable name autos.
autos=pd.read_csv("autos.csv",encoding="Latin-1")

In [2]:
#Explore key characteristics of autos dataframe

#Return an overview of al dtypes used in dataframe
autos.info()

#Return the dimensions of a dataframe
autos.shape


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

(50000, 20)

In [3]:
#View first 5 rows of dataframe
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Observations so far

The eBase Car Sales is a dataset which consists of 50000 rows and 20 columns. 5 out of 20 columns are integers. The other 15 columns are formatted as strings. 5 columns have cells containing null values. The header is in englisch, however the rest of the dataset is in German language. Furthermore, the header uses both small and capital letters.
The price column has a dollar sign, which should be the EURO sign I guess. Since it's a german dataset. 

# Clean Columns

In [4]:
#Existing column names
camelcase_header=autos.columns
camelcase_header

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
#Modified column names

#Modifying column names using function
def clean_col_labels(col):
    col=col.strip()
    col=col.replace("yearOfRegistration","registration_year")
    col=col.replace("monthOfRegistration","registration_month")
    col=col.replace("notRepairedDamage","unrepaired_damage")
    col=col.replace("dateCreated","ad_created")
    col=col.lower()
    return col

autos.columns=[clean_col_labels(i) for i in camelcase_header]

#Manually changing column names from camelcase to snakecase
snakecase_header=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns=snakecase_header

#Check modification
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [6]:
#Display current state of the autos dataframe after column name modification
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Initial Data Exploring and Cleaning

In [7]:
#Return descriptive statistics for all columns
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

# Cleaning actions
<b>Convert from object to integer</b>
<li>price                  50000 non-null object => remove dollar sign, no decimals
<li>odometer               50000 non-null object => remove km, no decimals

<b>Integer already</b>
<li>registration_year      50000 non-null int64 => weird years
<li>power_ps                50000 non-null int64 => weird ps
<li>registration_month     50000 non-null int64 => month=0???
<li>nr_of_pictures           50000 non-null int64 => all zeros DELETE
<li>postal_code             50000 non-null int64 => postal code with 4 digits

<b>Date Time Columns</b>
<li>date_crawled            50000 non-null object
<li>ad_created            50000 non-null object
<li>last_seen               50000 non-null object

<b>Object already</b>
<li>name                   50000 non-null object => brand,type,liter clean!
<li>seller                 50000 non-null object DELETE
<li>offer_type              50000 non-null object DELETE
<li>abtest                 50000 non-null object
<li>brand                  50000 non-null object => sonstige_autos 546/50000

<b>Null values</b>
<li>vehicle_type            44905 non-null object
<li>gearbox                47320 non-null object
<li>model                  47242 non-null object
<li>fuel_type               45518 non-null object
<li>unrepaired_damage      40171 non-null object


In [8]:
#Explore columns
#unique_values=autos["brand"].unique()
#count_values=autos["vehicle_type"].isnull().value_counts()
count_values=autos["price"].value_counts()
count_values


$0          1421
$500         781
$1,500       734
$2,500       643
$1,000       639
$1,200       639
$600         531
$800         498
$3,500       498
$2,000       460
$999         434
$750         433
$900         420
$650         419
$850         410
$700         395
$4,500       394
$300         384
$2,200       382
$950         379
$1,100       376
$1,300       371
$3,000       365
$550         356
$1,800       355
$5,500       340
$350         335
$1,250       335
$1,600       327
$1,999       322
            ... 
$9,970         1
$2,789         1
$9,130         1
$2,910         1
$299,000       1
$1,432         1
$35,700        1
$17,830        1
$3             1
$2,944         1
$19,100        1
$1,820         1
$11,989        1
$38,700        1
$33,980        1
$33,449        1
$27,299        1
$8,720         1
$7,820         1
$23,790        1
$1,494         1
$68,900        1
$6,155         1
$21,700        1
$24,699        1
$69,999        1
$2,459         1
$71,000       

In [9]:
#Converting a string column to Integer
autos["odometer"]=(autos["odometer"]
                            .str.replace("km","")
                            .str.replace(",","")
                            .astype(int)
                              )
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)

#Renaming an existing column
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos.rename({"price": "price_$"}, axis=1, inplace=True)

In [10]:
#Check conversion string to integer
#autos["price_$"].head()
count_values=autos["price_$"].value_counts()
count_values

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price_$, Length: 2357, dtype: int64

# Exploring the Odometer and Price Columns

In [11]:
#Explore the Odometer and Price Columns
#autos["odometer_km"].unique().shape
print(autos["odometer_km"].describe())
autos["price_$"].value_counts()
#check=autos["odometer_km"].sort_index(by, ascensding=True)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price_$, Length: 2357, dtype: int64

# Findings
<li>Odometer_km ranges from 5000 to 150000 and has 13 unique values. The km range from 10000 till 150000 correlates with the frequency. The higher the km, the more cars present in de dataframe. Since both extremes (5000 and 150000) are highly possible km numbers, I will leave it as it is.

<li>price_$ ranges from 0 to 151990 and has 2357 unique values. Since a price of 0 is not credible, I will treat it as an outlier and remove it from the autos dataframe. 

In [12]:
#Sort the counted values from the highest to lowest values
autos["price_$"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price_$, dtype: int64

In [13]:
#Sort the counted values from the lowest to highest values
autos["price_$"].value_counts().sort_index(ascending=True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price_$, dtype: int64

# Cleaning Price Column
There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [14]:
#Removing outliers and check the remaining values
autos=autos[autos["price_$"].between(1,350001)]
autos["price_$"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

# Exploring the date column

In [15]:
#Explore the date columns
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [16]:
#Returns object containing counts of unique values.

#autos['date_crawled'].str[:10].unique()
(autos['date_crawled']
     .str[:10] #select the first 10 characters
     .value_counts(normalize=True, dropna=False) #count unique values(percentages,including missing values)
     .sort_index(ascending=True) #rank by date in ascending order
     )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

date_crawled ranges from 5march16 till 7april16. The distribution of listed crawles is roughly even every day. 



In [17]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

ad_created ranges from 11jun15 till 7apr16. 

In [18]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

# Dealing with Incorrect Registration Year Data

In [19]:
#Explore highest and lowest acceptable values for the registratioin_year column
#autos["registration_year"].describe()

#(autos['registration_year']
#     .value_counts(normalize=True, dropna=False) #count unique values(percentages,including missing values)
#     .sort_index(ascending=True) #rank by date in ascending order
#     )

explore_registration_year_with_other_columns=autos.loc[autos["registration_year"]<=1900,"model"].value_counts(dropna=False)
print(explore_registration_year_with_other_columns)

#percentage of our data has invalid values in this column
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]



NaN       3
kaefer    1
andere    1
Name: model, dtype: int64


0.038793369710697

Not acceptable values
registration_year has a min of 1000 and a max of 9999 which is not likely. Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Both columns (ad_created and last_seen) ends at 7apr16. 
Given that this (registration_year not between 1900 and 2016) is less than 4% of our data, we will remove these rows.

In [20]:
#Remove rows
autos=autos[autos["registration_year"].between(1900,2017)]

#Check removal
(autos['registration_year']
     .value_counts(normalize=True, dropna=False) #count unique values(percentages,including missing values)
     .sort_index(ascending=True) #rank by date in ascending order
     )

1910    0.000104
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000083
1957    0.000042
1958    0.000083
1959    0.000125
1960    0.000478
1961    0.000125
1962    0.000083
1963    0.000166
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000541
1968    0.000541
          ...   
1988    0.002808
1989    0.003619
1990    0.007218
1991    0.007052
1992    0.007697
1993    0.008841
1994    0.013084
1995    0.025524
1996    0.028561
1997    0.040584
1998    0.049154
1999    0.060263
2000    0.065650
2001    0.054833
2002    0.051713
2003    0.056144
2004    0.056227
2005    0.061074
2006    0.055541
2007    0.047365
2008    0.046076
2009    0.043372
2010    0.033054
2011    0.033761
2012    0.027250
2013    0.016704
2014    0.013792
2015    0.0081

Most of the cars are registered between 1997 and 2009.

# Exploring Price by Brand

In [21]:
#Explore brand column

print(autos["brand"].describe())

(autos['brand']
     .value_counts(normalize=True, dropna=False) #count unique values(percentages,including missing values)
     #.sort_index(ascending=True) #rank by brand in ascending order
     )

count          48073
unique            40
top       volkswagen
freq           10220
Name: brand, dtype: object


volkswagen        0.212593
bmw               0.108814
opel              0.108377
mercedes_benz     0.095896
audi              0.086098
ford              0.069831
renault           0.047677
peugeot           0.029601
fiat              0.025836
seat              0.018618
skoda             0.016163
nissan            0.015310
mazda             0.015248
smart             0.014312
citroen           0.014104
toyota            0.012668
hyundai           0.009985
sonstige_autos    0.009631
volvo             0.009090
mini              0.008695
mitsubishi        0.008154
honda             0.007905
kia               0.007114
alfa_romeo        0.006636
porsche           0.005949
suzuki            0.005866
chevrolet         0.005637
chrysler          0.003515
dacia             0.002663
daihatsu          0.002496
jeep              0.002226
subaru            0.002101
land_rover        0.002059
saab              0.001623
daewoo            0.001560
jaguar            0.001519
trabant           0.001394
r

Brand data
<ol>
The brand column counts 40 unique car brands. Volkswagen is the most popular brand in this dataframe. There is a rest category which is called sonstige_autos representing almost 1%.

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [22]:
#Select brands to aggregate
brand_counts=autos['brand'].value_counts(normalize=True, dropna=False)
common_brands=brand_counts[brand_counts >.05].index
print(common_brands)

unique_brand=autos["brand"].unique()

#Select option
selected_brand=common_brands #or unique_brand

#Aggregate
aggregate_mean_price_per_brand={}

for i in selected_brand:
    mean_price_per_brand=autos.loc[autos["brand"]==i,"price_$"].mean()
    aggregate_mean_price_per_brand[i]=int(mean_price_per_brand)
aggregate_mean_price_per_brand


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


{'audi': 9239,
 'bmw': 8284,
 'ford': 3732,
 'mercedes_benz': 8528,
 'opel': 2952,
 'volkswagen': 5351}

Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

# Storing Aggregate Data in a DataFrame

In [23]:
#Select brands to aggregate
brand_counts=autos['brand'].value_counts(normalize=True, dropna=False)
common_brands=brand_counts[brand_counts >.05].index
print(common_brands)

unique_brand=autos["brand"].unique()

#Select option
selected_brand=common_brands #or unique_brand

#Aggregate
aggregate_mean_mileage_per_brand={}

for i in selected_brand:
    mean_mileage_per_brand=autos.loc[autos["brand"]==i,"odometer_km"].mean()
    aggregate_mean_mileage_per_brand[i]=int(mean_mileage_per_brand)
aggregate_mean_mileage_per_brand

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


{'audi': 129406,
 'bmw': 132666,
 'ford': 124255,
 'mercedes_benz': 130962,
 'opel': 129415,
 'volkswagen': 128928}

In [24]:
#Turn both dictionaries to series objects, using the series constructor
price_series=pd.Series(aggregate_mean_price_per_brand)
mileage_series=pd.Series(aggregate_mean_mileage_per_brand)

#Create a dataframe from the first series object using the dataframe constructor
df_price_mileage=pd.DataFrame(price_series,columns=["mean_price_per_brand"])

#Assign the other series as a new column in this dataframe
df_price_mileage["mean_mileage_per_brand"]=mileage_series

#Print the dataframe
df_price_mileage

mean_price_per_brand  mean_mileage_per_brand
audi                           9239                  129406
bmw                            8284                  132666
ford                           3732                  124255
mercedes_benz                  8528                  130962
opel                           2952                  129415
volkswagen                     5351                  128928

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

# Scrap Paper
screen_res = laptops["screen"].str.rsplit(n=1, expand=True)
screen_res.columns = ["A", "B"]
screen_res.loc[screen_res["B"].isnull(), "B"] = screen_res["A"]
laptops["screen_resolution"] = (screen_res["B"]
                                    .str.split(n=1,expand=True)
                                    .iloc[:,0]
                                    )
                                    
                                    
date_crawled_Y=(autos["date_crawled"]
                         .str[:10]
                         .str.split("-",expand=True)
                         .iloc[:,0]
                         .astype(int)
                        )

date_crawled_M=(autos["date_crawled"]
                         .str[:10]
                         .str.split("-",expand=True)
                         .iloc[:,1]
                         .astype(int)
                        )

date_crawled_D=(autos["date_crawled"]
                         .str[:10]
                         .str.split("-",expand=True)
                         .iloc[:,2]
                         .astype(int)
                        )

import datetime
autos["date_crawled"]=datetime.datetime(date_crawled_Y,date_crawled_M,date_crawled_D)


top_employer_by_country={}
unique_country=f500["country"].unique()
for i in unique_country:
    selected_countries=f500[f500["country"]==i]
    sorted_countries=selected_countries.sort_values(by=["employees"], ascending=False)
    company=sorted_countries.iloc[0,0]
    top_employer_by_country[i]=company